#!/bin/bash
# Variant_Calling
# Generated by dx-app-wizard.

main() {
    
    dx download "$tbam" -o ${caseid}.tumor.bam
    dx download "$reference" -o ref.tar.gz
    
    mkdir dnaref
    docker run -v ${PWD}:/data docker.io/goalconsortium/variantcalling:1.0.0 tar -I pigz -xvf ref.tar.gz --strip-components=1 -C dnaref
    
    panelopt=''
    if [ -n "$panel" ]
    then
        dx download "$panel" -o panel.tar.gz
        docker run -v ${PWD}:/data docker.io/goalconsortium/variantcalling:1.0.0 tar -I pigz -xvf panel.tar.gz
	capturebed=targetpanel.bed
	pon=mutect2.pon.vcf.gz
	panelopt="-b $capturebed"
    fi
    
    normopt=''
    if [ -n "$nbam" ]
    then
        dx download "$nbam" -o ${caseid}.normal.bam
        normopt=" -n ${caseid}.normal.bam"
    fi
    
    ponopt=''
    if [ -f "$pon" ]
    then
        ponopt="-q $pon"
    fi

    docker run -v ${PWD}:/data docker.io/goalconsortium/variantcalling:1.0.0 bash /seqprg/school/process_scripts/alignment/indexbams.sh
    
    vcfout=''
    vcfori=''
    outfile=''
    
    echo "$algo"
    
    for a in $algo
    do
        echo "Starting ${a}"
	vcfout+=" ${caseid}.${a}.vcf.gz"
	vcfori+=" ${caseid}.${a}.ori.vcf.gz"
	outfile+=".${a}"
	if [[ "${a}" == "fb" ]] || [[ "${a}" == "platypus" ]]
	then
	    docker run -v ${PWD}:/data docker.io/goalconsortium/variantcalling:1.0.0 bash /seqprg/school/process_scripts/variants/germline_vc.sh -r dnaref -p ${caseid} -a ${a} ${panelopt}
	    docker run -v ${PWD}:/data docker.io/goalconsortium/variantcalling:1.0.0 bash /seqprg/school/process_scripts/variants/uni_norm_annot.sh -g 'GRCh38.86' -r dnaref -p ${caseid}.${a} -v ${caseid}.${a}.vcf.gz
	    
	elif [[ "${a}" == "strelka2" ]] || [[ "${a}" == "mutect" ]] || [[ "${a}" == "shimmer" ]]
	then
	    if [[ -z "$nbam" ]]
	    then
		docker run -v ${PWD}:/data docker.io/goalconsortium/variantcalling:1.0.0 bash /seqprg/school/process_scripts/variants/germline_vc.sh -r dnaref -p ${caseid} -a ${a} ${panelopt} ${ponopt}
	    else
		docker run -v ${PWD}:/data docker.io/goalconsortium/variantcalling:1.0.0 bash /seqprg/school/process_scripts/variants/somatic_vc.sh -r dnaref -p ${caseid} -n ${caseid}.normal.bam -t ${caseid}.tumor.bam -a ${a} ${panelopt} ${ponopt}
	    fi
	    docker run -v ${PWD}:/data docker.io/goalconsortium/variantcalling:1.0.0 bash /seqprg/school/process_scripts/variants/uni_norm_annot.sh -g 'GRCh38.86' -r dnaref -p ${caseid}.${a} -v ${caseid}.${a}.vcf.gz
	else
	    echo "Incorrect algorithm selection. Please select 1 of the following algorithms: fb platypus strelka2 mutect shimmer"
	fi
    done
    
    tar cf ${caseid}${outfile}.vcfout.tar $vcfout
    tar cf ${caseid}${outfile}.ori.tar $vcfori
    gzip ${caseid}${outfile}.vcfout.tar
    gzip ${caseid}${outfile}.ori.tar
    
    
    vcf=$(dx upload ${caseid}${outfile}.vcfout.tar.gz --brief)
    ori=$(dx upload ${caseid}${outfile}.ori.tar.gz --brief)
    
    dx-jobutil-add-output vcf "$vcf" --class=file
    dx-jobutil-add-output ori "$ori" --class=file
}
